### PR TITLE
feat(query): Settings clause

### PR DESCRIPTION
**Pull request by TCeason**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

Settings clause

Syntax

```sql
settings (setting_name=value, ...) stmt
```

More info: https://github.com/databendlabs/databend/issues/16387

Fixes: https://github.com/databendlabs/databend/issues/16387

## Tests

- [ ] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/databendlabs/databend/16669)

### CODE REVIEW

File: tests/sqllogictests/suites/query/set.test [hunk 1]
```diff
@@ -64,3 +64,67 @@ select * from identifier($t);
...
+
+query TTTT
+select id, c from t union all select id, c from t1 order by id, c;
+----
+1 2022-02-01 16:00:00.000000
+1 2022-02-02 00:00:00.000000
+NULL 2022-02-02 00:00:00.000000
+
+query TTTT
+settings ((timezone, default_order_by_null)=('Asia/Shanghai', 'nulls_first')) select id, c from t union all select id, c from t1 order by id, c;
```
Comment (sundy-li) [inline tests/sqllogictests/suites/query/set.test:118]: I think we should use the following syntax:
```
SETTINGS ( parameter = value [ , parameter = value ] )  Statement
```

Comment (TCeason) [reply tests/sqllogictests/suites/query/set.test:118]: But now, in our set syntax is :

```sql
SET <var> = <expr>
SET ( <var> [ , <var> ... ] )  = ( <expr> [ , <expr> ... ] )
```

I think it's better to keep the two as consistent as possible

**Comment by TCeason** [comment]
cc @soyeric128 Doc needed. Settings is better than set_var hints. So we prefer to use `settings`.

File: src/query/ast/src/parser/statement.rs [hunk 2]
```diff
@@ -2782,6 +2795,46 @@ pub fn hint(i: Input) -> IResult<Hint> {
...
+            for (id, value) in query_setting {
+                ids.push(id);
+                values.push(value);
+            }
+            Settings {
+                set_type: SetType::SettingsQuery,
+                identifiers: ids,
+                values: SetValues::Expr(values.into_iter().map(|x| x.into()).collect()),
+            }
+            // let (ids, values): (Vec<Identifier>, Vec<Expr>) = query_setting
```
Comment (sundy-li) [inline src/query/ast/src/parser/statement.rs:2824]: remove these useless codes?

File: src/query/ast/src/ast/statements/statement.rs [hunk 3]
```diff
@@ -38,6 +39,10 @@ use crate::ast::Query;
 #[derive(Debug, Clone, PartialEq, Drive, DriveMut)]
 pub enum Statement {
     Query(Box<Query>),
+    QueryWithSetting {
```
Comment (sundy-li) [inline src/query/ast/src/ast/statements/statement.rs:42]: StatementWithSettings

File: src/query/sql/src/planner/binder/query_setting.rs [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+use databend_common_expression::Expr;
+use databend_common_functions::BUILTIN_FUNCTIONS;
+
+use crate::plans::Plan;
+use crate::BindContext;
+use crate::Binder;
+use crate::TypeChecker;
+
+impl Binder {
+    pub async fn bind_query_setting(
```
Comment (sundy-li) [inline src/query/sql/src/planner/binder/query_setting.rs:36]: bind_statment_settings

**Review by sundy-li — APPROVED** [review]
Others LGTM

File: src/query/ast/src/ast/statements/statement.rs (hunk 3 above)
Comment (TCeason) [reply src/query/ast/src/ast/statements/statement.rs]: Done

File: src/query/ast/src/parser/statement.rs (hunk 2 above)
Comment (TCeason) [reply src/query/ast/src/parser/statement.rs]: Done

File: src/query/sql/src/planner/binder/query_setting.rs (hunk 4 above)
Comment (TCeason) [reply src/query/sql/src/planner/binder/query_setting.rs]: Done

**Pull request merged**